test/test_connection_pool.rb

```diff
@@ -554,4 +554,50 @@ def test_stats_with_string_size
       assert_equal(1, pool.available)
     end
   end
+
+  def test_removing_and_replacing_connections
+    created = 0
+    pool = ConnectionPool.new(timeout:0, size: 1) do
+      created += 1
+      NetworkConnection.new
+    end
+
+    conn = pool.checkout
+    pool.checkin
+
+    pool.remove_connection(conn)
+    pool.with{  }
+
+    assert_equal 2, created
+  end
+
+  def test_reaping
+    created = 0
+    pool = ConnectionPool.new(timeout:0, size: 1, reap_after: 1, reaping_frequency: 1) do
+      created += 1
+      NetworkConnection.new
+    end
+
+    pool.with{}
+    sleep 2
```

Comment:
This isn't ideal, but I didn't want to introduce some mocking framework like timecop without feedback